src/shared/components/TokenAmountInput.tsx

```diff
@@ -46,42 +46,72 @@ export default function TokenAmountInput({
 }: {
   label?: string
   inputLabel: string
-  amount: string
+  amount: bigint | null
   tokenAddress: string
   disabled?: boolean
-  onChange: (value: string) => void
+  onChange: (value: bigint | null) => void
   onValidate?: (value: boolean) => void
 }) {
+  const [textAmount, setTextAmount] = useState("")
+
   const balance = useDappSelector((state) =>
     selectTokenBalanceByAddress(state, tokenAddress)
   )
   const symbol = useDappSelector((state) =>
     selectTokenSymbolByAddress(state, tokenAddress)
   )
 
-  const validate = (value: string) => {
-    const result = handleValidate(value, balance)
-    const hasError = "error" in result
+  const validate = useCallback(
+    (value: string) => {
+      const result = handleValidate(value, balance)
+      const hasError = "error" in result
 
-    onValidate?.(!hasError)
-    return result
-  }
+      onValidate?.(!hasError)
+      return result
+    },
+    [balance, onValidate]
+  )
+
+  useEffect(() => {
+    const textToBigIntAmount =
+      textAmount === "" ? null : userAmountToBigInt(textAmount) ?? 0n
+
+    const bigIntToTextAmount = bigIntToPreciseUserAmount(balance)
+
+    // As we may be loosing some precision, we need to compare the values.
+    // Clicking "Max" button may result in bigint that is too big to be
+    // represented as a float number. In this case we need to compare values to
+    // not override the external value that stores the bigint using greater precision.
+    if (textToBigIntAmount !== amount && textAmount !== bigIntToTextAmount) {
+      onChange(textToBigIntAmount)
+    }
+
+    // Make sure this is working only one way:
+    // from the text provided by input to the parent component
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [textAmount, onChange])
+
+  useEffect(() => {
+    // Allow clearing the input from parent componentthis should be the only case
+    // where parent component is allowed to change the value
+    if (amount === null) {
+      setTextAmount("")
+    }
+  }, [amount])
```

Comment:
I would turn it into a hook f.e. `useTextAmount` - it may be useful in the future

Reply:
Really? I think it is very specific to this component, don't know what component other than input would need it. I would for sure want to battle test it first before making it reusable 😅